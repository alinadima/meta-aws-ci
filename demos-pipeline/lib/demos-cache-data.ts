import * as cdk from 'aws-cdk-lib';
import * as efs from 'aws-cdk-lib/aws-efs';
import * as ec2 from 'aws-cdk-lib/aws-ec2';
import { Construct } from 'constructs';

/**
 * ...
 */
export interface DemosPipelineCacheProps extends cdk.StackProps {
    /** */
        readonly vpc: ec2.IVpc;
}
/**
 * ...
 */
export class DemosPipelineCache extends cdk.Stack {
    public readonly filesystem: efs.IFileSystem;
    public readonly accesspoint: efs.IAccessPoint;

    constructor(scope: Construct, id: string, props: DemosPipelineCacheProps) {
        super(scope, id, props);
        const filesystem = new efs.FileSystem(this, 'DemosPipelineFilesystem', {
            vpc: props.vpc,
        });

        this.accesspoint = filesystem.addAccessPoint('DemosPipelineAccessPoint', {
            path: '/data',
        });

        this.filesystem = filesystem;
    }
}
