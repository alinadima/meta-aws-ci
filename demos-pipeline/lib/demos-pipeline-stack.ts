import * as cdk from 'aws-cdk-lib';
import { Construct } from 'constructs';
import * as codepipeline from 'aws-cdk-lib/aws-codepipeline';
import * as codepipeline_actions from 'aws-cdk-lib/aws-codepipeline-actions';
import { BuildSpec, LinuxBuildImage, PipelineProject } from 'aws-cdk-lib/aws-codebuild';
import { IRepository } from 'aws-cdk-lib/aws-ecr';

/**
 *
 */
export interface DemosPipelineProps extends cdk.StackProps {
    readonly githubOrg?: string;
    readonly githubRepo?: string;
    readonly codestarConnectionArn: string;
    readonly imageRepo: IRepository;
    readonly imageTag?: string;
}

/**
 * The core pipeline stack.
 */
export class DemosPipelineStack extends cdk.Stack {
    constructor(scope: Construct, id: string, props: DemosPipelineProps) {
        super(scope, id, props);

        const sourceOutput = new codepipeline.Artifact();
        const sourceAction = new codepipeline_actions.CodeStarConnectionsSourceAction({
            output: sourceOutput,
            actionName: 'Demo-Source',
            repo: props.githubRepo ?? 'meta-aws-demos',
            owner: props.githubOrg ?? 'aws4embeddedlinux',
            connectionArn: props.codestarConnectionArn,
            codeBuildCloneOutput: true,
        });

        const project = new PipelineProject(this, 'DemoBuildProject', {
            buildSpec: BuildSpec.fromAsset('assets/demo/qemu/buildspec.yml'),
            environment: { buildImage: LinuxBuildImage.fromEcrRepository(props.imageRepo, props.imageTag) },
        });

        const buildAction = new codepipeline_actions.CodeBuildAction({
            input: sourceOutput,
            actionName: 'Demo-Build',
            project,
        });

        new codepipeline.Pipeline(this, 'DemoPipeline', {
            restartExecutionOnUpdate: true,
            stages: [
                {
                    stageName: 'Source',
                    actions: [sourceAction],
                },
                {
                    stageName: 'Build',
                    actions: [buildAction],
                },
            ],
        });
    }
}
