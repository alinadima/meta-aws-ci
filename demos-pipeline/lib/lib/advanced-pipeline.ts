import * as cdk from "aws-cdk-lib";
import { Construct } from "constructs";
import * as codepipeline from "aws-cdk-lib/aws-codepipeline";
import * as codepipeline_actions from "aws-cdk-lib/aws-codepipeline-actions";
import {
  BuildSpec,
  ComputeType,
  FileSystemLocation,
  LinuxBuildImage,
  PipelineProject,
} from "aws-cdk-lib/aws-codebuild";
import { IRepository } from "aws-cdk-lib/aws-ecr";
import * as efs from "aws-cdk-lib/aws-efs";
import {
  ISecurityGroup,
  IVpc,
  Peer,
  Port,
  SecurityGroup,
} from "aws-cdk-lib/aws-ec2";
import { Bucket } from "aws-cdk-lib/aws-s3";

/**
 * The device to build demos for.
 */
export enum AdvancedDeviceKind {
  /**  Qemu x86-64 */
  Qemu = "qemu",
}

/**
 * Properties to allow customizing the build.
 */
export interface AdvancedPipelineProps extends cdk.StackProps {
  /** GitHub Repository Owner */
  readonly githubOrg?: string;
  /** GitHub Repository Name. */
  readonly githubRepo?: string;
  /** GitHub Branch Name. */
  readonly githubBranch?: string;
  /** ECR Repository where the Build Host Image resides. */
  readonly imageRepo: IRepository;
  /** Tag for the Build Host Image */
  readonly imageTag?: string;
  /** VPC where the networking setup resides. */
  readonly vpc: IVpc;
  /** Demo Device to build for. */
  readonly device?: AdvancedDeviceKind;
}

/**
 * The stack demonstrating how to build a pipeline for meta-aws-demos
 */
export class AdvancedPipelineStack extends cdk.Stack {
  constructor(scope: Construct, id: string, props: AdvancedPipelineProps) {
    super(scope, id, props);
    /** Set up a default value for the demo BUILD_DEVICE. */
    const device = props.device ?? AdvancedDeviceKind.Qemu;

    /** Set up networking access and EFS FileSystems. */

    const projectSg = new SecurityGroup(this, "BuildProjectSecurityGroup", {
      vpc: props.vpc,
      description: "Security Group to allow attaching EFS",
    });
    projectSg.addIngressRule(
      Peer.ipv4(props.vpc.vpcCidrBlock),
      Port.tcp(2049),
      "NFS Mount Port"
    );

    const sstateFS = this.addFileSystem("SState", props.vpc, projectSg);
    const dlFS = this.addFileSystem("Downloads", props.vpc, projectSg);
    const tmpFS = this.addFileSystem("Temp", props.vpc, projectSg);

    const connectionArn = this.node.getContext("connectionarn");

    if (connectionArn === undefined) {
      throw new Error("No CodeStar Connection ARN provided in CDK Context!");
    }

    /** Create our CodePipeline Actions. */

    const sourceOutput = new codepipeline.Artifact();
    const sourceAction =
      new codepipeline_actions.CodeStarConnectionsSourceAction({
        output: sourceOutput,
        actionName: "Demo-Source",
        repo: props.githubRepo ?? "meta-aws-demos",
        branch: props.githubBranch ?? "master-next",
        owner: props.githubOrg ?? "aws4embeddedlinux",
        connectionArn,
        codeBuildCloneOutput: true,
      });

    const project = new PipelineProject(this, "DemoBuildProject", {
      buildSpec: BuildSpec.fromAsset(
        `assets/demo/${device}/build.buildspec.yml`
      ),
      environment: {
        computeType: ComputeType.X2_LARGE,
        buildImage: LinuxBuildImage.fromEcrRepository(
          props.imageRepo,
          props.imageTag
        ),
        privileged: true,
      },
      timeout: cdk.Duration.hours(4),
      vpc: props.vpc,
      securityGroups: [projectSg],
      fileSystemLocations: [
        FileSystemLocation.efs({
          identifier: "tmp_dir",
          location: tmpFS,
          mountPoint: "/build-output",
        }),
        FileSystemLocation.efs({
          identifier: "sstate_cache",
          location: sstateFS,
          mountPoint: "/sstate-cache",
        }),
        FileSystemLocation.efs({
          identifier: "dl_dir",
          location: dlFS,
          mountPoint: "/downloads",
        }),
      ],
    });

    const buildOutput = new codepipeline.Artifact();
    const buildAction = new codepipeline_actions.CodeBuildAction({
      input: sourceOutput,
      actionName: "Demo-Build",
      outputs: [buildOutput],
      project,
    });

    const artifactBucket = new Bucket(this, "DemoArtifact", {
      versioned: true,
      enforceSSL: true,
    });

    const artifactAction = new codepipeline_actions.S3DeployAction({
      actionName: "Demo-Artifact",
      input: buildOutput,
      bucket: artifactBucket,
    });

    /** Now create the actual Pipeline */
    new codepipeline.Pipeline(this, "DemoPipeline", {
      restartExecutionOnUpdate: true,
      stages: [
        {
          stageName: "Source",
          actions: [sourceAction],
        },
        {
          stageName: "Build",
          actions: [buildAction],
        },
        {
          stageName: "Artifact",
          actions: [artifactAction],
        },
      ],
    });
  }

  /**
   * Adds an EFS FileSystem to the VPC and SecurityGroup.
   *
   * @param name - A name to differentiate the filesystem.
   * @param vpc - The VPC the Filesystem resides in.
   * @param securityGroup - A SecurityGroup to allow access to the filesystem from.
   * @returns The filesystem location URL.
   *
   */
  private addFileSystem(
    name: string,
    vpc: IVpc,
    securityGroup: ISecurityGroup
  ): string {
    const fs = new efs.FileSystem(this, `DemoPipeline${name}Filesystem`, {
      vpc,
    });
    fs.connections.allowFrom(securityGroup, Port.tcp(2049));

    const fsId = fs.fileSystemId;
    const region = cdk.Stack.of(this).region;

    return `${fsId}.efs.${region}.amazonaws.com:/`;
  }
}
