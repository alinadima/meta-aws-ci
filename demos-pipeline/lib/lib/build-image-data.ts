import * as cdk from "aws-cdk-lib";
import { Construct } from "constructs";
import * as s3 from "aws-cdk-lib/aws-s3";
import { BucketDeployment, Source } from "aws-cdk-lib/aws-s3-deployment";
import * as iam from "aws-cdk-lib/aws-iam";

/**
 * Select options for the BuildImageDataStack.
 */
export interface BuildImageDataProps extends cdk.StackProps {
  /** The bucket name for image build sources. This must be globally unique. */
  readonly bucketName: string;
}

/**
 * Input (Source) data for our Build Image Pipeline.
 */
export class BuildImageDataStack extends cdk.Stack {
  /** The bucket where source for Build Images are. */
  public readonly bucket: s3.IBucket;

  constructor(scope: Construct, id: string, props: BuildImageDataProps) {
    super(scope, id, props);

    this.bucket = this.createDeploymentBucket(props.bucketName);
  }

  /**
   * Create a bucket and S3 deployment to this bucket.
   *
   * @param bucketName The name of the bucket. Must be globally unique.
   * @param env Environment passed to the stack.
   */
  private createDeploymentBucket(bucketName: string): s3.IBucket {
    // Create a bucket, then allow a deployment Lambda to upload to it.
    const dataBucket = new s3.Bucket(this, "BuildImageDataBucket", {
      bucketName,
      versioned: true,
      removalPolicy: cdk.RemovalPolicy.DESTROY,
      autoDeleteObjects: true,
    });

    const dataBucketDeploymentRole = new iam.Role(
      this,
      "BuildImageBucketRole",
      {
        assumedBy: new iam.ServicePrincipal("lambda.amazonaws.com"),
        managedPolicies: [
          iam.ManagedPolicy.fromAwsManagedPolicyName(
            "service-role/AWSLambdaBasicExecutionRole"
          ),
        ],
      }
    );

    const region = cdk.Stack.of(this).region;
    const account = cdk.Stack.of(this).account;

    dataBucketDeploymentRole.addToPolicy(
      new iam.PolicyStatement({
        actions: ["kms:Decrypt"],
        resources: [`arn:aws:kms:${region}:${account}:key/*`],
      })
    );

    new BucketDeployment(this, "BuildImageBucketDeployment", {
      // Note: Run `npm run zip-data` before deploying this stack!
      sources: [Source.asset("../lib/dist/assets/build-image")],
      destinationBucket: dataBucket,
      role: dataBucketDeploymentRole,
      extract: true,
    });

    return dataBucket;
  }
}
