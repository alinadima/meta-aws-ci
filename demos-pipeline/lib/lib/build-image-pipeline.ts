import * as cdk from "aws-cdk-lib";
import { Construct } from "constructs";
import * as codepipeline from "aws-cdk-lib/aws-codepipeline";
import * as codepipeline_actions from "aws-cdk-lib/aws-codepipeline-actions";
import * as codebuild from "aws-cdk-lib/aws-codebuild";
import { IRepository } from "aws-cdk-lib/aws-ecr";
import { IBucket } from "aws-cdk-lib/aws-s3";
import * as events from "aws-cdk-lib/aws-events";
import { CodePipeline } from "aws-cdk-lib/aws-events-targets";

/**
 * The type of Image to build on.
 */
export enum ImageKind {
  /** Ubuntu 22.04 (LTS) */
  Ubuntu22_04 = "ubuntu_22_04",
  /** Fedora 37 */
  Fedora37 = "fedora_37",
  /** Debian GNU/Linux 11.x (Bullseye) */
  Debian11 = "debian_11",
}

/**
 * Select options for the BuildImagePipelineStack.
 */
export interface BuildImagePipelineProps extends cdk.StackProps {
  /** The Image type to create. */
  readonly imageKind: ImageKind;
  /** The data bucket from the BuildImageDataStack */
  readonly dataBucket: IBucket;
  /** The ECR Repository to push to. */
  readonly repository: IRepository;
}

/**
 * The pipeline for building the CodeBuild Image used in other pipelines. This
 * will produce an image for an OS based on verified Yocto hosts.
 */
export class BuildImagePipelineStack extends cdk.Stack {
  constructor(scope: Construct, id: string, props: BuildImagePipelineProps) {
    super(scope, id, props);

    // Create a source action.
    const sourceOutput = new codepipeline.Artifact("BuildImageSource");
    const sourceAction = new codepipeline_actions.S3SourceAction({
      actionName: "Build-Image-Source",
      bucket: props.dataBucket,
      bucketKey: "data.zip", // TODO(glimsdal): Parameterize.
      output: sourceOutput,
    });

    // Create a build action.
    const buildImageProject = new codebuild.PipelineProject(
      this,
      "BuildImageProject",
      {
        buildSpec: codebuild.BuildSpec.fromSourceFilename(
          `${props.imageKind}/buildspec.yml`
        ),
        environment: {
          computeType: codebuild.ComputeType.LARGE,
          buildImage: codebuild.LinuxBuildImage.STANDARD_7_0,
          privileged: true,
        },
        environmentVariables: {
          ECR_REPOSITORY_URI: {
            type: codebuild.BuildEnvironmentVariableType.PLAINTEXT,
            value: props.repository.repositoryUri,
          },
          AWS_ACCOUNT_ID: {
            type: codebuild.BuildEnvironmentVariableType.PLAINTEXT,
            value: cdk.Stack.of(this).account,
          },
          AWS_DEFAULT_REGION: {
            type: codebuild.BuildEnvironmentVariableType.PLAINTEXT,
            value: cdk.Stack.of(this).region,
          },
          IMAGE_TAG: {
            type: codebuild.BuildEnvironmentVariableType.PLAINTEXT,
            value: props.imageKind,
          },
        },
      }
    );
    props.repository.grantPullPush(buildImageProject);

    const buildAction = new codepipeline_actions.CodeBuildAction({
      actionName: "Build",
      project: buildImageProject,
      input: sourceOutput,
    });

    const pipeline = new codepipeline.Pipeline(this, "BuildImagePipeline", {
      pipelineName: `${props.imageKind}BuildImagePipeline`,
      stages: [
        {
          stageName: "Source",
          actions: [sourceAction],
        },
        {
          stageName: "Build",
          actions: [buildAction],
        },
      ],
      restartExecutionOnUpdate: true,
    });

    // Run this pipeline weekly to update the image OS.
    const pipelineTarget = new CodePipeline(pipeline);
    new events.Rule(this, "WeeklySchedule", {
      schedule: events.Schedule.cron({
        weekDay: "Monday",
        minute: "0",
        hour: "6",
      }),
      targets: [pipelineTarget],
    });
  }
}
