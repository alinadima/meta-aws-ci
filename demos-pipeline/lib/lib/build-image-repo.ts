import * as cdk from "aws-cdk-lib";
import { Construct } from "constructs";
import * as ecr from "aws-cdk-lib/aws-ecr";

/**
 * The ECR Repository to store build host images.
 */
export class BuildImageRepoStack extends cdk.Stack {
  /** The respository to put the build host container in. */
  public readonly repository: ecr.IRepository;

  constructor(scope: Construct, id: string, props?: cdk.StackProps) {
    super(scope, id, props);

    this.repository = new ecr.Repository(this, "BuildImageRepo", {});
  }
}
