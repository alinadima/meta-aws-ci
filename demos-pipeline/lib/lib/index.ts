export * from './network';
export * from './build-image-data';
export * from './build-image-repo';
export * from './build-image-pipeline';
export * from './demo-pipeline';
