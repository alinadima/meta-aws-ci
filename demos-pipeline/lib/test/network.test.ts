import * as cdk from "aws-cdk-lib";
import { Template } from "aws-cdk-lib/assertions";
import { PipelineNetworkStack } from "../lib/network";

describe("Pipeline Networking", () => {
  const props = {
    env: { account: "111111111111", region: "eu-central-1" },
  };

  test("Snapshot", () => {
    const app = new cdk.App();
    const stack = new PipelineNetworkStack(app, props);
    const template = Template.fromStack(stack);
    expect(template).toMatchSnapshot();
  });
});
