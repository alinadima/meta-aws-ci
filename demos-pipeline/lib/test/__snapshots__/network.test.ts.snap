// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pipeline Networking Snapshot 1`] = `
{
  "Parameters": {
    "BootstrapVersion": {
      "Default": "/cdk-bootstrap/hnb659fds/version",
      "Description": "Version of the CDK Bootstrap resources in this environment, automatically retrieved from SSM Parameter Store. [cdk:skip]",
      "Type": "AWS::SSM::Parameter::Value<String>",
    },
  },
  "Resources": {
    "PipelineVpc0543904A": {
      "Properties": {
        "CidrBlock": "10.0.0.0/16",
        "EnableDnsHostnames": true,
        "EnableDnsSupport": true,
        "InstanceTenancy": "default",
        "Tags": [
          {
            "Key": "Name",
            "Value": "PipelineNetwork/PipelineVpc",
          },
        ],
      },
      "Type": "AWS::EC2::VPC",
    },
    "PipelineVpcIGW3FA4A524": {
      "Properties": {
        "Tags": [
          {
            "Key": "Name",
            "Value": "PipelineNetwork/PipelineVpc",
          },
        ],
      },
      "Type": "AWS::EC2::InternetGateway",
    },
    "PipelineVpcPrivateSubnet1DefaultRouteC225ACC6": {
      "Properties": {
        "DestinationCidrBlock": "0.0.0.0/0",
        "NatGatewayId": {
          "Ref": "PipelineVpcPublicSubnet1NATGatewayA4388274",
        },
        "RouteTableId": {
          "Ref": "PipelineVpcPrivateSubnet1RouteTable6D8B603D",
        },
      },
      "Type": "AWS::EC2::Route",
    },
    "PipelineVpcPrivateSubnet1RouteTable6D8B603D": {
      "Properties": {
        "Tags": [
          {
            "Key": "Name",
            "Value": "PipelineNetwork/PipelineVpc/PrivateSubnet1",
          },
        ],
        "VpcId": {
          "Ref": "PipelineVpc0543904A",
        },
      },
      "Type": "AWS::EC2::RouteTable",
    },
    "PipelineVpcPrivateSubnet1RouteTableAssociation791F1EF2": {
      "Properties": {
        "RouteTableId": {
          "Ref": "PipelineVpcPrivateSubnet1RouteTable6D8B603D",
        },
        "SubnetId": {
          "Ref": "PipelineVpcPrivateSubnet1Subnet442DCECC",
        },
      },
      "Type": "AWS::EC2::SubnetRouteTableAssociation",
    },
    "PipelineVpcPrivateSubnet1Subnet442DCECC": {
      "Properties": {
        "AvailabilityZone": "dummy1a",
        "CidrBlock": "10.0.96.0/19",
        "MapPublicIpOnLaunch": false,
        "Tags": [
          {
            "Key": "aws-cdk:subnet-name",
            "Value": "Private",
          },
          {
            "Key": "aws-cdk:subnet-type",
            "Value": "Private",
          },
          {
            "Key": "Name",
            "Value": "PipelineNetwork/PipelineVpc/PrivateSubnet1",
          },
        ],
        "VpcId": {
          "Ref": "PipelineVpc0543904A",
        },
      },
      "Type": "AWS::EC2::Subnet",
    },
    "PipelineVpcPrivateSubnet2DefaultRouteF3A9A6B9": {
      "Properties": {
        "DestinationCidrBlock": "0.0.0.0/0",
        "NatGatewayId": {
          "Ref": "PipelineVpcPublicSubnet2NATGateway5B528D21",
        },
        "RouteTableId": {
          "Ref": "PipelineVpcPrivateSubnet2RouteTable9A31913F",
        },
      },
      "Type": "AWS::EC2::Route",
    },
    "PipelineVpcPrivateSubnet2RouteTable9A31913F": {
      "Properties": {
        "Tags": [
          {
            "Key": "Name",
            "Value": "PipelineNetwork/PipelineVpc/PrivateSubnet2",
          },
        ],
        "VpcId": {
          "Ref": "PipelineVpc0543904A",
        },
      },
      "Type": "AWS::EC2::RouteTable",
    },
    "PipelineVpcPrivateSubnet2RouteTableAssociationC502A5A4": {
      "Properties": {
        "RouteTableId": {
          "Ref": "PipelineVpcPrivateSubnet2RouteTable9A31913F",
        },
        "SubnetId": {
          "Ref": "PipelineVpcPrivateSubnet2SubnetE21FED10",
        },
      },
      "Type": "AWS::EC2::SubnetRouteTableAssociation",
    },
    "PipelineVpcPrivateSubnet2SubnetE21FED10": {
      "Properties": {
        "AvailabilityZone": "dummy1b",
        "CidrBlock": "10.0.128.0/19",
        "MapPublicIpOnLaunch": false,
        "Tags": [
          {
            "Key": "aws-cdk:subnet-name",
            "Value": "Private",
          },
          {
            "Key": "aws-cdk:subnet-type",
            "Value": "Private",
          },
          {
            "Key": "Name",
            "Value": "PipelineNetwork/PipelineVpc/PrivateSubnet2",
          },
        ],
        "VpcId": {
          "Ref": "PipelineVpc0543904A",
        },
      },
      "Type": "AWS::EC2::Subnet",
    },
    "PipelineVpcPrivateSubnet3DefaultRoute8706A1DD": {
      "Properties": {
        "DestinationCidrBlock": "0.0.0.0/0",
        "NatGatewayId": {
          "Ref": "PipelineVpcPublicSubnet3NATGateway69775315",
        },
        "RouteTableId": {
          "Ref": "PipelineVpcPrivateSubnet3RouteTableBF7276B0",
        },
      },
      "Type": "AWS::EC2::Route",
    },
    "PipelineVpcPrivateSubnet3RouteTableAssociationEB416B63": {
      "Properties": {
        "RouteTableId": {
          "Ref": "PipelineVpcPrivateSubnet3RouteTableBF7276B0",
        },
        "SubnetId": {
          "Ref": "PipelineVpcPrivateSubnet3SubnetFCC0E247",
        },
      },
      "Type": "AWS::EC2::SubnetRouteTableAssociation",
    },
    "PipelineVpcPrivateSubnet3RouteTableBF7276B0": {
      "Properties": {
        "Tags": [
          {
            "Key": "Name",
            "Value": "PipelineNetwork/PipelineVpc/PrivateSubnet3",
          },
        ],
        "VpcId": {
          "Ref": "PipelineVpc0543904A",
        },
      },
      "Type": "AWS::EC2::RouteTable",
    },
    "PipelineVpcPrivateSubnet3SubnetFCC0E247": {
      "Properties": {
        "AvailabilityZone": "dummy1c",
        "CidrBlock": "10.0.160.0/19",
        "MapPublicIpOnLaunch": false,
        "Tags": [
          {
            "Key": "aws-cdk:subnet-name",
            "Value": "Private",
          },
          {
            "Key": "aws-cdk:subnet-type",
            "Value": "Private",
          },
          {
            "Key": "Name",
            "Value": "PipelineNetwork/PipelineVpc/PrivateSubnet3",
          },
        ],
        "VpcId": {
          "Ref": "PipelineVpc0543904A",
        },
      },
      "Type": "AWS::EC2::Subnet",
    },
    "PipelineVpcPublicSubnet1DefaultRoute44F3E91D": {
      "DependsOn": [
        "PipelineVpcVPCGW3256101F",
      ],
      "Properties": {
        "DestinationCidrBlock": "0.0.0.0/0",
        "GatewayId": {
          "Ref": "PipelineVpcIGW3FA4A524",
        },
        "RouteTableId": {
          "Ref": "PipelineVpcPublicSubnet1RouteTableE9A67515",
        },
      },
      "Type": "AWS::EC2::Route",
    },
    "PipelineVpcPublicSubnet1EIPBD0800F9": {
      "Properties": {
        "Domain": "vpc",
        "Tags": [
          {
            "Key": "Name",
            "Value": "PipelineNetwork/PipelineVpc/PublicSubnet1",
          },
        ],
      },
      "Type": "AWS::EC2::EIP",
    },
    "PipelineVpcPublicSubnet1NATGatewayA4388274": {
      "DependsOn": [
        "PipelineVpcPublicSubnet1DefaultRoute44F3E91D",
        "PipelineVpcPublicSubnet1RouteTableAssociation6D13736B",
      ],
      "Properties": {
        "AllocationId": {
          "Fn::GetAtt": [
            "PipelineVpcPublicSubnet1EIPBD0800F9",
            "AllocationId",
          ],
        },
        "SubnetId": {
          "Ref": "PipelineVpcPublicSubnet1Subnet26FF83E2",
        },
        "Tags": [
          {
            "Key": "Name",
            "Value": "PipelineNetwork/PipelineVpc/PublicSubnet1",
          },
        ],
      },
      "Type": "AWS::EC2::NatGateway",
    },
    "PipelineVpcPublicSubnet1RouteTableAssociation6D13736B": {
      "Properties": {
        "RouteTableId": {
          "Ref": "PipelineVpcPublicSubnet1RouteTableE9A67515",
        },
        "SubnetId": {
          "Ref": "PipelineVpcPublicSubnet1Subnet26FF83E2",
        },
      },
      "Type": "AWS::EC2::SubnetRouteTableAssociation",
    },
    "PipelineVpcPublicSubnet1RouteTableE9A67515": {
      "Properties": {
        "Tags": [
          {
            "Key": "Name",
            "Value": "PipelineNetwork/PipelineVpc/PublicSubnet1",
          },
        ],
        "VpcId": {
          "Ref": "PipelineVpc0543904A",
        },
      },
      "Type": "AWS::EC2::RouteTable",
    },
    "PipelineVpcPublicSubnet1Subnet26FF83E2": {
      "Properties": {
        "AvailabilityZone": "dummy1a",
        "CidrBlock": "10.0.0.0/19",
        "MapPublicIpOnLaunch": true,
        "Tags": [
          {
            "Key": "aws-cdk:subnet-name",
            "Value": "Public",
          },
          {
            "Key": "aws-cdk:subnet-type",
            "Value": "Public",
          },
          {
            "Key": "Name",
            "Value": "PipelineNetwork/PipelineVpc/PublicSubnet1",
          },
        ],
        "VpcId": {
          "Ref": "PipelineVpc0543904A",
        },
      },
      "Type": "AWS::EC2::Subnet",
    },
    "PipelineVpcPublicSubnet2DefaultRoute04C861A0": {
      "DependsOn": [
        "PipelineVpcVPCGW3256101F",
      ],
      "Properties": {
        "DestinationCidrBlock": "0.0.0.0/0",
        "GatewayId": {
          "Ref": "PipelineVpcIGW3FA4A524",
        },
        "RouteTableId": {
          "Ref": "PipelineVpcPublicSubnet2RouteTable5219ED4D",
        },
      },
      "Type": "AWS::EC2::Route",
    },
    "PipelineVpcPublicSubnet2EIPB92B2C07": {
      "Properties": {
        "Domain": "vpc",
        "Tags": [
          {
            "Key": "Name",
            "Value": "PipelineNetwork/PipelineVpc/PublicSubnet2",
          },
        ],
      },
      "Type": "AWS::EC2::EIP",
    },
    "PipelineVpcPublicSubnet2NATGateway5B528D21": {
      "DependsOn": [
        "PipelineVpcPublicSubnet2DefaultRoute04C861A0",
        "PipelineVpcPublicSubnet2RouteTableAssociation54D39738",
      ],
      "Properties": {
        "AllocationId": {
          "Fn::GetAtt": [
            "PipelineVpcPublicSubnet2EIPB92B2C07",
            "AllocationId",
          ],
        },
        "SubnetId": {
          "Ref": "PipelineVpcPublicSubnet2Subnet64F58E18",
        },
        "Tags": [
          {
            "Key": "Name",
            "Value": "PipelineNetwork/PipelineVpc/PublicSubnet2",
          },
        ],
      },
      "Type": "AWS::EC2::NatGateway",
    },
    "PipelineVpcPublicSubnet2RouteTable5219ED4D": {
      "Properties": {
        "Tags": [
          {
            "Key": "Name",
            "Value": "PipelineNetwork/PipelineVpc/PublicSubnet2",
          },
        ],
        "VpcId": {
          "Ref": "PipelineVpc0543904A",
        },
      },
      "Type": "AWS::EC2::RouteTable",
    },
    "PipelineVpcPublicSubnet2RouteTableAssociation54D39738": {
      "Properties": {
        "RouteTableId": {
          "Ref": "PipelineVpcPublicSubnet2RouteTable5219ED4D",
        },
        "SubnetId": {
          "Ref": "PipelineVpcPublicSubnet2Subnet64F58E18",
        },
      },
      "Type": "AWS::EC2::SubnetRouteTableAssociation",
    },
    "PipelineVpcPublicSubnet2Subnet64F58E18": {
      "Properties": {
        "AvailabilityZone": "dummy1b",
        "CidrBlock": "10.0.32.0/19",
        "MapPublicIpOnLaunch": true,
        "Tags": [
          {
            "Key": "aws-cdk:subnet-name",
            "Value": "Public",
          },
          {
            "Key": "aws-cdk:subnet-type",
            "Value": "Public",
          },
          {
            "Key": "Name",
            "Value": "PipelineNetwork/PipelineVpc/PublicSubnet2",
          },
        ],
        "VpcId": {
          "Ref": "PipelineVpc0543904A",
        },
      },
      "Type": "AWS::EC2::Subnet",
    },
    "PipelineVpcPublicSubnet3DefaultRouteDD586919": {
      "DependsOn": [
        "PipelineVpcVPCGW3256101F",
      ],
      "Properties": {
        "DestinationCidrBlock": "0.0.0.0/0",
        "GatewayId": {
          "Ref": "PipelineVpcIGW3FA4A524",
        },
        "RouteTableId": {
          "Ref": "PipelineVpcPublicSubnet3RouteTableD6F00678",
        },
      },
      "Type": "AWS::EC2::Route",
    },
    "PipelineVpcPublicSubnet3EIPA3CA7521": {
      "Properties": {
        "Domain": "vpc",
        "Tags": [
          {
            "Key": "Name",
            "Value": "PipelineNetwork/PipelineVpc/PublicSubnet3",
          },
        ],
      },
      "Type": "AWS::EC2::EIP",
    },
    "PipelineVpcPublicSubnet3NATGateway69775315": {
      "DependsOn": [
        "PipelineVpcPublicSubnet3DefaultRouteDD586919",
        "PipelineVpcPublicSubnet3RouteTableAssociation163F9BEA",
      ],
      "Properties": {
        "AllocationId": {
          "Fn::GetAtt": [
            "PipelineVpcPublicSubnet3EIPA3CA7521",
            "AllocationId",
          ],
        },
        "SubnetId": {
          "Ref": "PipelineVpcPublicSubnet3SubnetE08894FF",
        },
        "Tags": [
          {
            "Key": "Name",
            "Value": "PipelineNetwork/PipelineVpc/PublicSubnet3",
          },
        ],
      },
      "Type": "AWS::EC2::NatGateway",
    },
    "PipelineVpcPublicSubnet3RouteTableAssociation163F9BEA": {
      "Properties": {
        "RouteTableId": {
          "Ref": "PipelineVpcPublicSubnet3RouteTableD6F00678",
        },
        "SubnetId": {
          "Ref": "PipelineVpcPublicSubnet3SubnetE08894FF",
        },
      },
      "Type": "AWS::EC2::SubnetRouteTableAssociation",
    },
    "PipelineVpcPublicSubnet3RouteTableD6F00678": {
      "Properties": {
        "Tags": [
          {
            "Key": "Name",
            "Value": "PipelineNetwork/PipelineVpc/PublicSubnet3",
          },
        ],
        "VpcId": {
          "Ref": "PipelineVpc0543904A",
        },
      },
      "Type": "AWS::EC2::RouteTable",
    },
    "PipelineVpcPublicSubnet3SubnetE08894FF": {
      "Properties": {
        "AvailabilityZone": "dummy1c",
        "CidrBlock": "10.0.64.0/19",
        "MapPublicIpOnLaunch": true,
        "Tags": [
          {
            "Key": "aws-cdk:subnet-name",
            "Value": "Public",
          },
          {
            "Key": "aws-cdk:subnet-type",
            "Value": "Public",
          },
          {
            "Key": "Name",
            "Value": "PipelineNetwork/PipelineVpc/PublicSubnet3",
          },
        ],
        "VpcId": {
          "Ref": "PipelineVpc0543904A",
        },
      },
      "Type": "AWS::EC2::Subnet",
    },
    "PipelineVpcVPCGW3256101F": {
      "Properties": {
        "InternetGatewayId": {
          "Ref": "PipelineVpcIGW3FA4A524",
        },
        "VpcId": {
          "Ref": "PipelineVpc0543904A",
        },
      },
      "Type": "AWS::EC2::VPCGatewayAttachment",
    },
  },
  "Rules": {
    "CheckBootstrapVersion": {
      "Assertions": [
        {
          "Assert": {
            "Fn::Not": [
              {
                "Fn::Contains": [
                  [
                    "1",
                    "2",
                    "3",
                    "4",
                    "5",
                  ],
                  {
                    "Ref": "BootstrapVersion",
                  },
                ],
              },
            ],
          },
          "AssertDescription": "CDK bootstrap stack version 6 required. Please run 'cdk bootstrap' with a recent version of the CDK CLI.",
        },
      ],
    },
  },
}
`;
